frontend/app-development/layout/SettingsModalButton/SettingsModal/components/Tabs/AccessControlTab/AccessControlTab.tsx

```diff
@@ -31,32 +41,89 @@ export type AccessControlTabProps = {
 
-  const handleChange = (newPartyTypes: string[], currentPartyTypesAllowed: PartyTypesAllowed) => {
-    const newPartyTypesAllowed = { ...currentPartyTypesAllowed };
+  const handleAllowedPartyTypeChange = (
+    newPartyTypes: string[],
+    currentPartyTypesAllowed: PartyTypesAllowed,
+    checkboxValue: string,
+  ) => {
+    const updatedPartyTypesAllowed = { ...currentPartyTypesAllowed };
+    updatedPartyTypesAllowed[checkboxValue] = !currentPartyTypesAllowed[checkboxValue];
 
-    Object.keys(currentPartyTypesAllowed).forEach((key) => {
-      newPartyTypesAllowed[key] = newPartyTypes.includes(key);
-    });
-    updateAppMetadataMutation({ ...appMetadata, partyTypesAllowed: newPartyTypesAllowed });
-  };
+    const updatedCheckboxes = Object.values(updatedPartyTypesAllowed).filter((value) => value);
 
-  const displayCheckboxes = () => {
-    return getPartyTypesAllowedOptions().map((option) => (
-      <Checkbox value={option.value} key={option.value} size='small'>
-        {t(option.label)}
-      </Checkbox>
-    ));
+    if (updatedCheckboxes.length === 0) {
+      modalRef.current?.showModal();
+      return;
+    }
+
+    try {
+      updateAppMetadataMutation({
+        ...appMetadata,
+        partyTypesAllowed: updatedPartyTypesAllowed,
+      });
+      refetchAppMetadata();
+    } catch (error) {
+      console.error('Error updating app metadata:', error);
+    }
   };
 
+  const isAnyCheckboxChecked = appMetadata
+    ? Object.values(appMetadata.partyTypesAllowed).some((value) => value)
+    : false;
+
+  const areAllCheckboxesChecked = appMetadata
+    ? Object.values(appMetadata.partyTypesAllowed).every((value) => value)
+    : false;
+
+  const tableHeaderCheckboxState = isAnyCheckboxChecked
+    ? areAllCheckboxesChecked
+      ? CheckboxState.Checked
+      : CheckboxState.Indeterminate
+    : CheckboxState.Unchecked;
+
+  const handleTableHeaderCheckboxChange = () => {
+    if (appMetadata) {
+      const updatedPartyTypesAllowed = { ...appMetadata.partyTypesAllowed };
+      const newValue = !areAllCheckboxesChecked;
+      Object.keys(updatedPartyTypesAllowed).forEach((key) => {
+        updatedPartyTypesAllowed[key] = newValue;
+      });
+      try {
+        updateAppMetadataMutation({
+          ...appMetadata,
+          partyTypesAllowed: updatedPartyTypesAllowed,
+        });
+        refetchAppMetadata();
+      } catch (error) {
+        console.error('Error updating app metadata:', error);
+      }
+    }
+  };
+  const renderModal = () => {
+    return (
+      <Modal ref={modalRef}>
+        <Modal.Content className={classes.modalContent}>
+          {t('settings_modal.access_control_tab_option_choose_type_modal_message')}
+        </Modal.Content>
+        <Modal.Footer>
+          <Button variant='secondary' onClick={() => modalRef.current?.close()}>
+            {t('general.close')}
+          </Button>
+        </Modal.Footer>
+      </Modal>
+    );
+  };
```

Comment:
Can this be separated to its own component? 😊

Reply:
fixed